xbmc/games/controllers/types/ControllerNode.h

```diff
@@ -0,0 +1,105 @@
+/*
+ *  Copyright (C) 2017-2021 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+#include "games/controllers/ControllerTypes.h"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace KODI
+{
+namespace GAME
+{
+class CControllerHub;
+
+/*!
+ * \brief Node in the controller tree
+ *
+ * The node identifies the controller profile, and optionally the available
+ * controller ports.
+ */
+class CControllerNode
+{
+public:
+  CControllerNode();
+  CControllerNode(const CControllerNode& other) { *this = other; }
+  CControllerNode(CControllerNode&& other) = default;
+  CControllerNode& operator=(const CControllerNode& rhs);
+  ~CControllerNode();
+
+  void Clear();
+
+  /*!
+   * \brief Controller profile of this code
+   *
+   * \return Controller profile, or empty if this node is invalid
+   *
+   * \sa IsValid()
+   */
+  const ControllerPtr& Controller() const { return m_controller; }
```

Comment:
Do you want to return a `const std::shared_ptr<CController>` or a `std::shared_ptr<const CController>`?

Reply:
Ideally the second but it seemed like a big change. Think we should adapt the rest of the code?

Reply:
Yeah I know. Ever since I realized the difference I always have two `using`s for `FooPtr` and `FooConstPtr` and when you use `auto` wherever you retrieve these pointers it's not that bad. AFAIK we have this "flaw" everywhere in the code base so no big deal if we add another one if it's too much hassle to fix.